design-proposals/cpu-hotplug.md

```diff
@@ -0,0 +1,146 @@
+While this design is focused on CPU hotplug, it aims to provide a generic API design for resource hotplug that will be reusable for things like memory.
+
+## Motivation
+CPU hotplug is now a common virtualization feature, but KubeVirt lacks support for it.
+Live vertical scaling of VM resources in general can be very useful in a variety of scenarios.
+
+## Goals
+- Hotplug and hot-unplug of CPU resources from VMs should be achievable by editing the VM object itself (no sub-resources)
+- Underlying technology used for hotplug should be independent of the VM API, meaning we should be able to swap out how hotplug is technically performed in the future without impacting API compatibility
+- Implementation should be achievable today, with Kubernetes APIs that are at least in beta. Unfortunately, at the time of writing, the Kubernetes vertical pod scaling API is only alpha
+
+## Non Goals
+- Revising existing hotplug mechanisms like Volume hotplug
+- Dynamically applying affinity rules to VMIs
+
+## Definition of Users
+A user with (at a minimum) create, update, and patch access for VM objects within a namespace.
+
+## User Stories
+As a VM user, I would like to dynamically increase or decrease the number of CPUs used by a VM without requiring the VM to restart.
+
+## Repos
+kubevirt/kubevirt
+
+# Design
+In VM objects, the existing CPU `sockets` field will now be dynamic, provided that it has been declared as such.
+Declaring the field as dynamic will be done by adding a `cpu` entry to a new section under the VM spec, called `liveUpdateFeatures`.
+Under the `cpu` entry, it will be possible to define a maximum number of sockets. That number is needed by LibVirt and will default to 4 times the initial number of sockets if not set.
+
+Increasing the number of CPU sockets will not only add CPUs to the guest but also increase the CPU resources available to the virt-launcher pod, via a migration (more on that in the Implementation section).
+The way LibVirt handles CPU hotplug is by exposing the maximum number of CPUs to the guest and turning off the unused CPUs. In this document, when we refer to "adding" CPUs to the guest, we're really just turning them on. This is why we need to define a maximum number of sockets in advance.
+
+## Drawbacks / Limitations
+- Since CPU hot-(un)plug involves a live migration, VMs will have to be `LiveMigratable` to enable CPU hotplug
+- Each disabled vCPU, so `(maxSockets - sockets) * cores * thread`, consumes 8MiB of overhead memory
+- This feature will be incompatible with CPU requests/limits, at least initially, since those values need to dynamically scale according to the current number of enabled CPUs
+
+## API Examples
+
+### Spec API
+
+```
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  labels:
+    kubevirt.io/vm: vm-cirros
+  name: vm-cirros
+spec:
+  running: false
+  liveUpdateFeatures:
+  - cpu
+      maximumSockets: 4
```

Comment:
This approach is not intended to live-update VMI resources directly?
Thus to hotplug CPU you always have to have the above VM resource, am I right?

Reply:
> Is that list with maps? I think it's better to make map of maps

Right, this has to be a map of maps, will fix.

> However, personally I don't like to have parameters like maximumSockets outside of domain section

Since `maximumSockets` is an attribute of the CPU live update feature, and since it doesn't make sense when the feature is not enabled, I thought having is there made the most sense. We could move it to `domain` and ensure the live update feature is enabled in a webhook, but I'm not sure it adds any value...

> This approach is not intended to live-update VMI resources directly?
> Thus to hotplug CPU you always have to have the above VM resource, am I right?

Live updates features would indeed require VMs and not just VMIs. Technically, final users of KubeVirt should only ever work with VMs, VMIs being just intermediate objects that they should not be concerned with. In reality though, VMIs do appear to be quite popular!
That being said, having the `liveUpdateFeatures` and dynamic `sockets` in boths VMs and VMIs feels like a bit of a headache to reconcile...